eo-maven-plugin/src/test/java/org/eolang/maven/optimization/OptCachedTest.java

```diff
@@ -0,0 +1,113 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.eolang.maven.optimization;
+
+import com.jcabi.xml.XML;
+import com.jcabi.xml.XMLDocument;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.cactoos.io.ResourceOf;
+import org.cactoos.io.UncheckedInput;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
+import org.hamcrest.io.FileMatchers;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+/**
+ * Test case for {@link org.eolang.maven.optimization.OptCached}.
+ * @since 0.28.12
+ */
+class OptCachedTest {
+    @Test
+    void optimizesIfXmlAlreadyInCache(final @TempDir Path tmp) {
+        final XML program = OptCachedTest.program();
+        MatcherAssert.assertThat(
+            new OptCached(path -> program, tmp).apply(OptCachedTest.save(tmp, program)),
+            Matchers.equalTo(program)
+        );
+    }
+
+    @Test
+    void optimizesIfXmlIsAbsentInCache(final @TempDir Path tmp) {
+        final XML program = OptCachedTest.program();
+        final Path cache = tmp.resolve("cache");
+        final XML res = new OptCached(path -> program, cache)
+            .apply(OptCachedTest.save(tmp, program));
+        MatcherAssert.assertThat(
+            res,
+            Matchers.equalTo(program)
+        );
+        MatcherAssert.assertThat(
+            cache.resolve("main.xmir").toFile(),
+            FileMatchers.anExistingFile()
+        );
+        MatcherAssert.assertThat(
+            res, Matchers.equalTo(program)
+        );
+    }
+
+    @Test
+    void throwsExceptionIfCantFindProgram(final @TempDir Path tmp) {
+        Assertions.assertThrows(
+            IllegalStateException.class,
+            () -> new OptCached(p -> OptCachedTest.program(), tmp.resolve("cache"))
+                .apply(tmp.resolve("non-existent"))
+        );
+    }
+
+    /**
+     * Save XML program with hardcoded name to a temp directory.
+     * @param tmp Temporary test directory.
+     * @param xml XML program.
+     * @return Path to saved program.
+     */
+    private static Path save(final Path tmp, final XML xml) {
+        final Path res = tmp.resolve("main.xmir");
+        try {
+            Files.write(res, xml.toString().getBytes(StandardCharsets.UTF_8));
+            return res;
+        } catch (final IOException ex) {
+            throw new IllegalStateException(String.format("Can't save XML to '%s'", res), ex);
+        }
+    }
+
+    /**
+     * Get parsed program from resources.
+     * @return XML representation of program.
+     */
+    private static XML program() {
+        final String path = "org/eolang/maven/optimize/main.xml";
+        try {
+            return new XMLDocument(
+                new UncheckedInput(new ResourceOf(path)).stream()
```

Comment:
@volodya-lombrozo it's a bad idea to call `.stream()` and not close it later. Don't do it. Instead, use `XMLDocument(path)`.

Reply:
@yegor256 I don't wanna create and save intermediate file in order to use `XMLDocument(path)`, I wanna load xml right from resources, by that reason I've used the next code:
```java
 private static XML program() {
        return new XMLDocument(
            new UncheckedBytes(
                new BytesOf(
                    new ResourceOf("org/eolang/maven/optimize/main.xml")
                )
            ).asBytes()
        );
    }
```